src/core/shaderlib/lighting/lighting.js

```diff
@@ -20,9 +20,131 @@
 
 import lightingShader from './lighting.glsl';
 import project from '../project/project';
+import {COORDINATE_SYSTEM} from '../../lib/constants';
+import vec4_transformMat4 from 'gl-vec4/transformMat4';
+import vec3_sub from 'gl-vec3/subtract';
 
 export default {
   name: 'lighting',
   dependencies: [project],
-  vs: lightingShader
+  vs: lightingShader,
+  getUniforms
 };
+
+const INITIAL_MODULE_OPTIONS = {};
+
+const DEFAULT_LIGHTS_POSITION = [-122.45, 37.75, 8000];
+const DEFAULT_LIGHTS_STRENGTH = [2.0, 0.0];
+const DEFAULT_AMBIENT_RATIO = 0.05;
+const DEFAULT_DIFFUSE_RATIO = 0.6;
+const DEFAULT_SPECULAR_RATIO = 0.8;
+
+function getUniforms(opts = INITIAL_MODULE_OPTIONS) {
+  if (!opts.lightSettings) {
+    return {};
+  }
+
+  const {
+    numberOfLights = 1,
+
+    lightsPosition = DEFAULT_LIGHTS_POSITION,
+    lightsStrength = DEFAULT_LIGHTS_STRENGTH,
+    coordinateSystem = COORDINATE_SYSTEM.LNGLAT,
+    coordinateOrigin = [0, 0, 0],
+    modelMatrix = null,
+
+    ambientRatio = DEFAULT_AMBIENT_RATIO,
+    diffuseRatio = DEFAULT_DIFFUSE_RATIO,
+    specularRatio = DEFAULT_SPECULAR_RATIO
+  } = opts.lightSettings;
+
+  const projectionParameters = {
+    viewport: opts.viewport,
+    coordinateSystem,
+    coordinateOrigin,
+    layerCoordinateSystem: opts.coordinateSystem,
+    layerCoordinateOrigin: opts.coordinateOrigin,
+    modelMatrix
+  };
+
+  const lightsPositionWorld = [];
+  for (let i = 0; i < numberOfLights; i++) {
+    const position = preProject(lightsPosition.slice(i * 3, i * 3 + 3), projectionParameters);
+
+    lightsPositionWorld[i * 3] = position[0];
+    lightsPositionWorld[i * 3 + 1] = position[1];
+    lightsPositionWorld[i * 3 + 2] = position[2];
+  }
+
+  return {
+    lighting_lightPositions: lightsPositionWorld,
+    lighting_lightStrengths: lightsStrength,
+    lighting_ambientRatio: ambientRatio,
+    lighting_diffuseRatio: diffuseRatio,
+    lighting_specularRatio: specularRatio,
+    lighting_numberOfLights: numberOfLights
+  };
+}
+
+/* Projection utils */
+
+function lngLatZToWorldPosition(lngLatZ, viewport) {
+  const [X, Y] = viewport.projectFlat(lngLatZ);
+  const Z = (lngLatZ[2] || 0) * viewport.scale;
+  return [X, Y, Z];
+}
+
+function preProject(
+  position,
+  {
+    viewport,
+    coordinateSystem,
+    coordinateOrigin,
+    modelMatrix,
+    layerCoordinateSystem,
+    layerCoordinateOrigin
+  }
+) {
+  let [x, y, z] = position;
+  let worldPosition;
+
+  if (modelMatrix) {
+    [x, y, z] = vec4_transformMat4([], [x, y, z, 1.0], modelMatrix);
+  }
+
+  // pre-project light coordinates
+  switch (coordinateSystem) {
+    case COORDINATE_SYSTEM.LNGLAT:
+      worldPosition = lngLatZToWorldPosition([x, y, z], viewport);
+      break;
+
+    case COORDINATE_SYSTEM.LNGLAT_OFFSETS:
```

Comment:
I think this switch statement should be handled by code in `Viewports` or `project` module. Otherwise we need to update dependent modules like this whenever we update the set of supported coordinate system.

Reply:
Moving this into the `project` module makes sense, since it is essentially the CPU equivalent of the `project_position` function in glsl.

Reply:
In fact, we are no longer depending on any shader function from the `project` module.